mne/beamformer/_dics.py

```diff
@@ -1,32 +1,452 @@
+            'subject' : {string}
+                The subject ID.
+
+    See Also
+    --------
+    apply_dics, dics, dics_epochs
+
+    Notes
+    -----
+    For more information about ``real_filter``, see the
+    `supplemental information <http://www.cell.com/cms/attachment/616681/4982593/mmc1.pdf>`_
+    from [2]_.
+
+    References
+    ----------
+    .. [1] Gross et al. Dynamic imaging of coherent sources: Studying neural
+           interactions in the human brain. PNAS (2001) vol. 98 (2) pp. 694-699
+    .. [2] Hipp JF, Engel AK, Siegel M (2011) Oscillatory Synchronization
+           in Large-Scale Cortical Networks Predicts Perception.
+           Neuron 69:387-396.
+    """  # noqa: E501
+    allowed_ori = [None, 'normal', 'max-power']
+    if pick_ori not in allowed_ori:
+        raise ValueError('"pick_ori" should be one of %s.' % allowed_ori)
+
+    frequencies = [np.mean(freq_bin) for freq_bin in csd.frequencies]
+    n_freqs = len(frequencies)
+    n_orient = forward['sol']['ncol'] // forward['nsource']
+
+    picks = _setup_picks(info=info, forward=forward)
+    _, ch_names, proj, vertices, G = _prepare_beamformer_input(
+        info, forward, label, picks=picks, pick_ori=pick_ori
+    )
+    csd_picks = [csd.names.index(ch) for ch in ch_names]
+
+    n_sources = G.shape[1] // n_orient
+
+    # Normalize the leadfield for depth weighting. This also ensures that
+    # G @ G.T == I, which simplifies some of the equations.
+    G = G / np.linalg.norm(G, axis=0)
+
+    logger.info('Computing DICS spatial filters...')
+    Ws = []
+    for i, freq in enumerate(frequencies):
+        if n_freqs > 1:
+            logger.info('    computing DICS spatial filter at %sHz (%d/%d)' %
+                        (freq, i + 1, n_freqs))
+
+        Cm = csd.get_matrix(index=i)
+
+        if real_filter:
+            Cm = Cm.real
+
+        # Ensure the CSD is in the same order as the leadfield
+        Cm = Cm[csd_picks, :][:, csd_picks]
+
+        # Tikhonov regularization using reg parameter to control for
+        # trade-off between spatial resolution and noise sensitivity
+        # eq. 25 in Gross and Ioannides, 1999 Phys. Med. Biol. 44 2081
+        Cm_inv, _ = _reg_pinv(Cm, reg, rcond='auto')
+
+        # Compute spatial filters
+        W = np.dot(G.T, Cm_inv)
+
+        for k in range(n_sources):
+            Wk = W[n_orient * k: n_orient * k + n_orient]
+            Gk = G[:, n_orient * k: n_orient * k + n_orient]
+
+            # Compute power at the source
+            Ck = np.dot(Wk, Gk)
+
+            # Normalize the spatial filters
+            if Wk.ndim == 2 and len(Wk) > 1:
+                # Free source orientation
+                # Officially, the formula calls for inv(Ck) @ Wk.
+                # However, since Ck is most likely rank deficient, the
+                # following is more stable in practice.
+                Wk /= np.diag(Ck)[:, np.newaxis]
```

Comment:
I think the weight normalization should be optional, too.

Reply:
Really? Again, can you give an example where you would want to turn it off?

Reply:
For example, you might want to correct for the depth bias by solely normalizing the leadfield (or, as mentioned above, the other way around). 
Also I think that normalization requires a different orientation selection formula (based on Sekihara & Nagarajan, 2008, p. 40).

Reply:
I'm going to read Sekihara & Nagarajan and consult with Jan Kujala and get back to you on this. At the present time, I would advise anyone who wants to do DICS powermapping to follow Kujala's pipeline, which means normalizing both the leadfield and the weights. Until I know of scenarios where I would honestly advice someone to do things differently, I'm hesitant to build in the switches to do so. This may very well be knowledge lacking from my end and I will do my best to remedy this. But "it *could* be done differently" or "fieldtrip does it differently" are not sufficient reasons for me to implement the switches in the present PR (someone may in the future though). I'm definitely going to look into the orientation selection formula. Stay tuned!

Reply:
Here are some examples of how lead field normalization tends to fail for us, and I believe similar failures are why it is no longer popular in the beamformer community. Any lead field normalization indeed cancels out if applied to contrasting conditions or to a weight-normalized beamformer, so this tends to mask its potential drawbacks. (But, even in that case, applying a normalizer that is later cancelled out is needless computation.)

We test much of our code internally based on responses to a full field flash, so we expect a large area of primary visual cortex to produce evoked high-frequency oscillations (105-145 Hz in this case). Our pipeline is based on FieldTrip, and I have personally contributed some of the underlying code and validated the formulas involved against both references (e.g. Sekihara & Nagarajan) and performance with real data. These are time domain examples, but it will also apply to DICS.

"Lead field normalization" often means that the columns of the lead field are normalized and so if sometimes referred to as "column normalization". With our full field flash data, this is the peak activation that is returned:
![colnorm](https://user-images.githubusercontent.com/17592830/33734981-457c8682-db8e-11e7-87b6-3e5ccf4f4cc4.png)

If you "expect" you know where your source is, as we do in this case, you might be able to recover something by navigating to the proper structure:
![colnorm2](https://user-images.githubusercontent.com/17592830/33735010-58e8bda8-db8e-11e7-9012-3a17e81195b0.png)

For sensory responses such as this, we prefer to use the "Neural Activity Index", which is a weight-normalized beamformer and is equivalent to the unit-noise-gain beamformer with a noise scaling factor. The performance there is more what we expect, with a central peak of activation along the calcarine amongst a larger cortical patch (due to the full-field stimulus):
![nai](https://user-images.githubusercontent.com/17592830/33735327-463a5256-db8f-11e7-845c-c7314784f7f0.png)

For completeness, normalizing the lead field with the Frobenius norm of the matrix (not column-wise) looks like this:
![normfro](https://user-images.githubusercontent.com/17592830/33735382-691979c8-db8f-11e7-9015-17bcf68ad723.png)

Again navigating to the expected area will yield something more familiar:
![normfro2](https://user-images.githubusercontent.com/17592830/33735395-6f5e3d5a-db8f-11e7-9b42-9b756553351f.png)

So, I strongly advocate that lead field normalization be an option and not hard-coded. :-)

Reply:
@sarangnemo Thanks for this nice explanation. And thank you @britta-wstnr for the 'heads up'. I'm now convinced we do need the switches. Do we have any advice to give to the users on when to normalize and when not to?

Reply:
@wmvanvliet Normalization is not useful when contrasting conditions, since it will cancel out anyway. (I believe that actually covers most use cases of DICS.) It makes sense only when there is no explicit contrast (e.g., evaluating a simple sensory response or epileptic activity). The weightnorm/NAI approach constitutes an explicit comparison to (theoretical) background noise, and works well in our experience.

The lead field normalization approaches attempt to do something similar, but does not control varying noise gains across the brain, so the behavior is IMO less stable. I also suspect (in retrospect) that it behaves especially poorly for BEM-based lead fields, since the 'normalizer' value can vary drastically across the brain for such head models, especially near the tissue boundaries.